extensions/libevent.c

```
@@ -0,0 +1,318 @@
+
+static void*
+run_event_loop (void *p)
+{
+  int microsec = 0;
+  struct timeval tv;
+
+  struct loop_data *data = p;
+
+  if (data->timeout < 0)
+    microsec = -1;
+  else if (data->timeout >= 0)
+    {
+      microsec = data->timeout / time_units_per_microsec;
+      tv.tv_sec = 0;
+      tv.tv_usec = microsec;
+    }
+
+  if (microsec >= 0)
+    event_base_loopexit (data->base, &tv);
+
+  event_base_loop (data->base, EVLOOP_ONCE);
+
+  return NULL;
+}
+
+static SCM
+scm_primitive_event_loop (SCM lst, SCM wakefd, SCM wokefd, SCM timeout)
+#define FUNC_NAME "primitive-event-loop"
+{
+  int c_wakefd;
+  int c_wokefd;
+  int result = 0;
+  int64_t c_timeout;
+
+  c_wakefd = scm_to_int (wakefd);
+  c_wokefd = scm_to_int (wokefd);
+  c_timeout = scm_to_int64 (timeout);
+
+  struct libevt_data *libevt =
+    (struct libevt_data *) scm_to_pointer (scm_list_ref (lst, scm_from_int (0)));
+  struct wait_data *data =
+    (struct wait_data *) scm_to_pointer (scm_list_ref (lst, scm_from_int (1)));
+
+  if (data == NULL)
+    scm_wrong_type_arg_msg ("event-loop", 1,
+                            scm_list_ref (lst, scm_from_int (1)),
+                            "event wait_data is null");
+
+  if (c_timeout != 0 && scm_c_prepare_to_wait_on_fd (c_wakefd))
+    data->rv = 0;
+  else
+    {
+      struct loop_data loop_data = { libevt->base, c_timeout };
+
+      if (c_timeout != 0)
+        scm_without_guile (run_event_loop, &loop_data);
+      else
+        run_event_loop (&loop_data);
+
+      if (c_timeout != 0)
+        scm_c_wait_finished ();
+
+      for (int i = 0; i < data->rv; i++)
+        {
+          if (data->events[i].fd == c_wokefd)
+            {
+              char zeroes[32];
+              /* Remove wake fd from result set.  */
+              data->rv--;
+              memmove (data->events + i,
+                       data->events + i + 1,
+                       (data->rv - i) * sizeof (struct event_data));
+              /* Drain fd and ignore errors. */
```

Maybe there is a good reason why errors are ignored, but it isn't explained.

---

With the comment I added I believe it's clear now why we ignore errors. This is the same as `epoll` code.